### PR TITLE
[MIN-156] Added transfer_to_liquidation_pool function for liquidation-pools

### PR DESCRIPTION
**Pull request by ales-by**
**Description:**

Added update_balance function for liquidation-pools

**Changes in storage or API:**
Added update_balance() as a wrapper for orml

### CODE REVIEW

File: pallets/liquidation-pools/src/lib.rs [hunk 1]
```diff
@@ -321,6 +323,22 @@ pub mod module {
...
 		}
+
+        #[pallet::weight(T::LiquidationPoolsWeightInfo::balance_liquidation_pools())] //FIXME
+		#[transactional]
+        pub fn update_balance(
+            origin: OriginFor<T>,
+            currency_id: CurrencyId, 
+            by_amount: Amount,
+        ) -> DispatchResultWithPostInfo {
+            let _ = ensure_none(origin)?;
```
Comment (Apollo50) [inline pallets/liquidation-pools/src/lib.rs]: You have to use `T::UpdateOrigin::ensure_origin(origin)?;` instead. It's utilitilies function which could be used only by root or member of Minterest council.

We use` ensure_none` for functions that can be called internally ( by OffchainWorker e.g ).

File: pallets/liquidation-pools/src/lib.rs [hunk 2]
```diff
@@ -122,6 +122,8 @@ pub mod module {
 		BalanceRatioChanged(CurrencyId, Rate),
 		///  Maximum ideal balance has been successfully changed: \[pool_id, new_threshold_value\]
 		MaxIdealBalanceChanged(CurrencyId, Option<Balance>),
+        ///  Balance has been successfully updated
```
Comment (Apollo50) [inline pallets/liquidation-pools/src/lib.rs]: Better to add attributs used in event to the comment.

File: pallets/liquidation-pools/src/lib.rs [hunk 3]
```diff
@@ -321,6 +323,22 @@ pub mod module {
 			Self::deposit_event(Event::LiquidationPoolsBalanced);
 			Ok(().into())
 		}
+
+        #[pallet::weight(T::LiquidationPoolsWeightInfo::balance_liquidation_pools())] //FIXME
+		#[transactional]
+        pub fn update_balance(
```
Comment (Apollo50) [inline pallets/liquidation-pools/src/lib.rs]: It would be good to add function description.

**Review by Apollo50 — COMMENTED** [review]
Would be just as cool to add test for new ex.

File: pallets/liquidation-pools/src/lib.rs (hunk 1 above)
Comment (ales-by) [reply pallets/liquidation-pools/src/lib.rs]: fixed

**Review by lut1k — DISMISSED** [review]

File: pallets/liquidation-pools/src/tests.rs [hunk 4]
```diff
@@ -195,3 +195,54 @@ fn calculate_ideal_balance_should_work() {
...
+		.execute_with(|| {
+			// Check that update balance work correctly with addition
+			// Liquidity pool value: 500000
+			// by_amount 10000
+			// Expected balance 510000
+			let by_amount = 10000;
+			assert_eq!(
+				TestLiquidationPools::update_balance(admin(), DOT, by_amount),
+				Ok(().into())
+			);
```
Comment (lut1k) [inline pallets/liquidation-pools/src/tests.rs:213]: better: `assert_ok!()'

File: pallets/liquidation-pools/src/tests.rs [hunk 5]
```diff
@@ -195,3 +195,54 @@ fn calculate_ideal_balance_should_work() {
...
+			assert_eq!(TestLiquidationPools::get_pool_available_liquidity(DOT), 510000);
+			// Check that update balance work correctly with substraction
+			// Liquidity pool value: 510000
+			// by_amount -10000
+			// Expected balance 500000
+			let by_amount = -10000;
+			assert_eq!(
+				TestLiquidationPools::update_balance(admin(), DOT, by_amount),
+				Ok(().into())
+			);
```
Comment (lut1k) [inline pallets/liquidation-pools/src/tests.rs:227]: same, see above

**Review by lut1k — APPROVED** [review]

**Review by lut1k — DISMISSED** [review]

File: pallets/liquidation-pools/src/lib.rs [hunk 6]
```diff
@@ -326,6 +333,25 @@ pub mod module {
 			Self::deposit_event(Event::LiquidationPoolsBalanced);
 			Ok(().into())
 		}
+
+		/// Updates balance of liquidation pool
+		/// - `pool_id`: PoolID to update
+		/// - `by_amount`: value to change the balance. If positive do add, else do remove.
+		#[pallet::weight(10_000)] // Stub, function is for dev purposes only.
```
Comment (DenisRomanovsky) [inline pallets/liquidation-pools/src/lib.rs:340]: This function should be available for Admins on prod, so it is not for dev purposes only. It'll be a part of launching pools flow.

File: pallets/liquidation-pools/src/lib.rs [hunk 7]
```diff
@@ -326,6 +333,25 @@ pub mod module {
...
+		#[transactional]
+		pub fn update_balance(
+			origin: OriginFor<T>,
+			pool_id: CurrencyId,
+			by_amount: Amount,
+		) -> DispatchResultWithPostInfo {
+			let _ = T::UpdateOrigin::ensure_origin(origin)?;
+			let module_account_id = Self::pools_account_id();
+
+			T::MultiCurrency::update_balance(pool_id, &module_account_id, by_amount)?;
```
Comment (DenisRomanovsky) [inline pallets/liquidation-pools/src/lib.rs:350]: Is there a check somewhere that the by_amount is positive?  Can we add it please?

File: pallets/liquidation-pools/src/lib.rs [hunk 8]
```diff
@@ -326,6 +333,25 @@ pub mod module {
...
+
+		/// Updates balance of liquidation pool
+		/// - `pool_id`: PoolID to update
+		/// - `by_amount`: value to change the balance. If positive do add, else do remove.
+		#[pallet::weight(10_000)] // Stub, function is for dev purposes only.
+		#[transactional]
+		pub fn update_balance(
+			origin: OriginFor<T>,
+			pool_id: CurrencyId,
+			by_amount: Amount,
```
Comment (DenisRomanovsky) [inline pallets/liquidation-pools/src/lib.rs:345]: We usually omit grammar suffix like this. 
```suggestion
			amount: Amount,
```

File: pallets/liquidation-pools/src/tests.rs [hunk 9]
```diff
@@ -195,3 +195,48 @@ fn calculate_ideal_balance_should_work() {
...
+			// by_amount 10000
+			// Expected balance 510000
+			let by_amount = 10000;
+			assert_ok!(TestLiquidationPools::update_balance(admin(), DOT, by_amount));
+
+			let expected_event = Event::liquidation_pools(crate::Event::BalanceUpdated(DOT, by_amount));
+			assert!(System::events().iter().any(|record| record.event == expected_event));
+
+			assert_eq!(TestLiquidationPools::get_pool_available_liquidity(DOT), 510000);
+			// Check that update balance work correctly with substraction
```
Comment (DenisRomanovsky) [inline pallets/liquidation-pools/src/tests.rs:216]: Yeah, this should not work. We already have a QA version of this operation. Your extrinsic will be connected to the real wallet and work in prod. So we don't want someone to exploit the API.

File: pallets/liquidation-pools/src/lib.rs (hunk 6 above)
Comment (Apollo50) [reply pallets/liquidation-pools/src/lib.rs:340]: I don't think so )

File: pallets/liquidation-pools/src/lib.rs (hunk 7 above)
Comment (ales-by) [reply pallets/liquidation-pools/src/lib.rs:350]: Well, I added a wrapper based on original orml functionality, and it allows use of negative values. I can add a check, if we need, not a problem.

File: pallets/liquidation-pools/src/tests.rs (hunk 9 above)
Comment (ales-by) [reply pallets/liquidation-pools/src/tests.rs]: fixed

File: pallets/liquidation-pools/src/tests.rs [hunk 10]
```diff
@@ -195,3 +195,23 @@ fn calculate_ideal_balance_should_work() {
...
 }
+#[test]
+fn transfer_to_liquidation_pool_should_work() {
+	ExternalityBuilder::default()
+		.liquidity_pool_balance(DOT, 500_000)
+		.user_balance(ADMIN, DOT, 20_000)
+		.build()
+		.execute_with(|| {
+			let who = ensure_signed(admin());
+			// Check that trasfer to liquidation pool worfs correctly
```
Comment (DenisRomanovsky) [inline pallets/liquidation-pools/src/tests.rs:206]:
```suggestion
			// Check that transfer to liquidation pool works correctly
```

**Review by DenisRomanovsky — DISMISSED** [review]

File: pallets/liquidation-pools/src/lib.rs [hunk 11]
```diff
@@ -112,6 +112,10 @@ pub mod module {
 		PoolNotFound,
 		/// Not enough liquidation pool balance.
 		NotEnoughBalance,
+
+		NotEnoughLiquidityAvailable,
```
Comment (Apollo50) [inline pallets/liquidation-pools/src/lib.rs:116]: Add comments please

File: pallets/liquidation-pools/src/tests.rs [hunk 12]
```diff
@@ -195,3 +195,23 @@ fn calculate_ideal_balance_should_work() {
 			);
 		});
 }
+#[test]
+fn transfer_to_liquidation_pool_should_work() {
```
Comment (Apollo50) [inline pallets/liquidation-pools/src/tests.rs:199]: Add please handling for each errors are possible.

**Review by lut1k — APPROVED** [review]

**Review by Apollo50 — APPROVED** [review]

**Pull request merged**